exonum-java-binding/doc/Migration_guide_0.9.md

```diff
@@ -0,0 +1,179 @@
+using the light client.
+
+[service-id-lc-operation]: todo
+
+## Update the Service
+
+The [`Service`][service] implementation must be updated to support multiple instances;
+new initialization and reconfiguration procedures.
+
+1. Remove the hardcoded service id and name. Add a constructor parameter 
+`ServiceInstanceSpec instanceSpec` which provides instance name
+and id.
+    - If your service extends
+    [`com.exonum.binding.core.service.AbstractService`][abstract-service] —
+    pass the `instanceSpec` to the superclass constructor and use provided `AbstractService#getId`,
+    `getName`, and `getInstanceSpec` methods
+    - If it implements `Service` directly — save the needed information to fields.
+
+```java
+public final class FooService extends AbstractService {
+  @Inject
+  public FooService(ServiceInstanceSpec instanceSpec) {
+    super(instanceSpec);
+  }
+}
+```
+
+2. If the service requires initialization — migrate to the new 
+[initialization mechanism][service-initialize]. 
+
+3. If the service needs reconfiguration support: implement [`Configurable`][configurable] interface.
+Reconfiguration shall be performed through the new [supervisor][supervisor] service which
+replaces the configuration service.
+
+[service]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/service/Service.html
+[service-initialize]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/service/Service.html#initialize-com.exonum.binding.core.storage.database.Fork-com.exonum.binding.core.service.Configuration-
+[configurable]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/service/Configurable.html
+[supervisor]: todo
+
+### Update the Schema
+
+#### Isolate Collections via Namespaces
+
+Use a unique for each service instance namespace for service collections (e.g., a service name,
+or an id).
+
+#### Update ProofMaps
+
+The default [`ProofMap`][proof-map] implementation has been changed to hash user keys to produce an internal key.
+That allows _any_ type to be used as a key; but adds an extra hashing operation for every insert
+operation and changes how proofs need to be verified.
+
+In some cases (rarer than is usually considered), it might be needed to **not** hash the keys
+(use them as _internal_ keys). Such `ProofMap` variant is still supported, see if your
+service implementation fits the [*requirements*][proof-map-non-hashing] to keep using it.
+If it doesn't — please migrate your schema to use the new default one.
+
+[proof-map]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/storage/indices/ProofMapIndexProxy.html
+[proof-map-non-hashing]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/storage/indices/ProofMapIndexProxy.html#key-hashing
+
+#### Disable Proofs
+
+Proofs are temporarily disabled. Creating proofs is _not_ supported in this release.
```

Comment:
Will you get any error if you try to get the prof from, for example, ListProofIndex? If yes mb it's good to mention it?

Reply:
Doesn't 'disabled and not supported' imply 'won't work, don't use'? I think we will `@Deprecate` the proof creating methods for this release (to keep tests for now, because we will need them later, but disallow usages); maybe — replace implementations with an exception. Shall we specify the exact behaviour here?

Reply:
>  Shall we specify the exact behavior here?

it's not required, just curious